react/features/connection-indicator/components/AbstractConnectionIndicator.js

```diff
@@ -0,0 +1,179 @@
+// @flow
+
+import { Component } from 'react';
+
+import statsEmitter from '../statsEmitter';
+
+declare var interfaceConfig: Object;
+
+/**
+ * The connection quality percentage that must be reached to be considered of
+ * good quality and can result in the connection indicator being hidden.
+ *
+ * @type {number}
+ */
+export const INDICATOR_DISPLAY_THRESHOLD = 30;
+
+/**
+ * The type of the React {@code Component} props of {@link ConnectionIndicator}.
+ */
+export type Props = {
+
+    /**
+     * The ID of the participant associated with the displayed connection indication and
+     * stats.
+     */
+    participantId: string
+};
+
+/**
+ * The type of the React {@code Component} state of {@link ConnectionIndicator}.
+ */
+export type State = {
+
+    /**
+     * The timeout for automatically hiding the indicator.
+     */
+    autoHideTimeout: ?TimeoutID,
+
+    /**
+     * Whether or not a CSS class should be applied to the root for hiding the
+     * connection indicator. By default the indicator should start out hidden
+     * because the current connection status is not known at mount.
+     */
+    showIndicator: boolean,
+
+    /**
+     * Cache of the stats received from subscribing to stats emitting. The keys
+     * should be the name of the stat. With each stat update, updates stats are
+     * mixed in with cached stats and a new stats object is set in state.
+     */
+    stats: Object
+};
+
+/**
+ * Implements a React {@link Component} which displays the current connection
+ * quality percentage and has a popover to show more detailed connection stats.
+ *
+ * @extends {Component}
+ */
+export default class AbstractConnectionIndicator<P: Props, S: State> extends Component<P, S> {
+    /**
+     * Initializes a new {@code ConnectionIndicator} instance.
+     *
+     * @param {P} props - The read-only properties with which the new
+     * instance is to be initialized.
+     */
+    constructor(props: P) {
+        super(props);
+
+        // Bind event handlers so they are only bound once for every instance.
+        this._onStatsUpdated = this._onStatsUpdated.bind(this);
+    }
+
+    /**
+     * Starts listening for stat updates.
+     *
+     * @inheritdoc
+     * returns {void}
+     */
+    componentDidMount() {
+        statsEmitter.subscribeToClientStats(
+            this.props.participantId, this._onStatsUpdated);
+    }
+
+    /**
+     * Updates which user's stats are being listened to.
+     *
+     * @inheritdoc
+     * returns {void}
+     */
+    componentDidUpdate(prevProps: Props) {
```

Comment:
why not set the participant ID as the key, so the component is recreated if it changes and we save having to handle updates?

Reply:
Probably true, but I have some concerns here:
1. I think recreating the component is more expensive than handling updates
2. I didn't spend too much time searching, but the docs I found about keys are mainly about lists and arrays. It never mentions or states that in case of a generic component (that is not a list element) a key change would result the component to be properly destroyed and a new created (although probably this is the case). Can you back this with some doc? If so, I'm happy to change it.

Reply:
The code is not in a hot-path, so I'd rather have it simpler than more complex for marginal performance gains.

Here is an explanation of `key`, the official docs a bit unwieldy: https://medium.com/@albertogasparin/forcing-state-reset-on-a-react-component-by-using-the-key-prop-14b36cd7448e

Reply:
Quote from this article: ```On real word applications, you should limit non-list components with keys and avoid setting it on top level ones, as it might be the cause of (hard to spot) performance issues.```

I really don't want to do this. For me, performance is before code complexity.

Reply:
I don't enjoy performance claims without numbers to back it up but whatever.